clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -915,25 +915,25 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
 
   ModulePassManager MPM;
 
-  // FIXME: Change this when -fno-sycl-early-optimizations is not tied to
-  // -disable-llvm-passes.
-  if (CodeGenOpts.DisableLLVMPasses && LangOpts.SYCLIsDevice)
+  if (LangOpts.SYCLIsDevice && !CodeGenOpts.DisableSYCLEarlyOpts)
     MPM.addPass(SYCLPropagateAspectsUsagePass());
 
   if (!CodeGenOpts.DisableLLVMPasses) {
+
     // Map our optimization levels into one of the distinct levels used to
     // configure the pipeline.
     OptimizationLevel Level = mapToLevel(CodeGenOpts);
 
-    if (LangOpts.SYCLIsDevice)
+    if (LangOpts.SYCLIsDevice && !CodeGenOpts.DisableSYCLEarlyOpts)
       PB.registerPipelineStartEPCallback(
           [&](ModulePassManager &MPM, OptimizationLevel Level) {
             MPM.addPass(ESIMDVerifierPass(LangOpts.SYCLESIMDForceStatelessMem));
             MPM.addPass(SYCLPropagateAspectsUsagePass());
           });
 
     // Add the InferAddressSpaces pass for all the SPIR[V] targets
-    if (TargetTriple.isSPIR() || TargetTriple.isSPIRV()) {
+    if ((TargetTriple.isSPIR() || TargetTriple.isSPIRV()) &&
+        !CodeGenOpts.DisableSYCLEarlyOpts) {
```

Comment:
This change is not needed, I believe, because `InferAddressSpacePass` is a functional pass